_posts/2016-12-05-haskell-package-checklist.md

```diff
@@ -0,0 +1,109 @@
+  - For example, generally avoid `lens`.
+  - Similarly, avoid having too many dependencies.
+- Include `extra-source-files`
+  - If it’s necessary in order to build, put it in here.
+  - Keep it as small as possible.
+  - Also include things for tests and benchmarks.
+- Keep executables small
+  - Define the executable in your library and re-export it.
+  - This allows others to use your executable’s behavior from Haskell.
+  - Avoid re-defining main because it causes problems with GHCi.
+- Avoid package warnings
+  - Stack prints these out when you run `stack sdist`.
+  - Some of the warnings, like category and description, are annoying.
+  - Generally using `--pvp-bounds=both` is good.
+- Build `-Wall` clean
+  - GHC can find all kinds of problems but doesn’t by default.
+  - Few false positives, generally helps you write better code.
+  - You can ignore specific warnings with `-fno-warn-whatever`.
+- Put everything in `source/`
+  - This separates Haskell files from package metadata.
+  - Makes it easy to search and script Haskell files.
+  - Looks like `source/library/Foo.hs`, `source/executables/Main.hs`, `source/tests/Main.hs`, `source/benchmarks/Main.hs`, and so on.
+- Match module and package names
+  - Module names use `CamelCase`, package names use `kebab-case`.
+  - If you package is named `foo-bar`, you should have a top-level module called `FooBar`.
+  - Not `Data.FooBar` or `Text.ParserCombinators.FooBar`.
+- Require one import
+  - Users should be able to hit the ground running with `import FooBar`.
+  - If necessary, re-export functions from other packages.
+  - Optimize for qualified imports, so don’t be afraid to take common names.
+- Document with examples
+  - Types are good, but they’re not documentation.
+  - The same is true for laws.
+  - Usually functions are added to solve a problem. Show that problem in the documentation as an example.
+- Expose implementation details
+  - People will use your package in ways you didn’t think of.
+  - Make everything public, but not necessarily published.
+  - Use `Internal` module names to signal that they’re private.
+- Follow most HLint suggestions
+  - Overall a great tool that helps you write better software.
+  - Some suggestions aren’t worth following.
+  - In particular the re-export shortcut and anything involving operators.
+- Format code with `hindent`
+  - This frees you up to never think about formatting again.
+  - Improving `hindent` improves formatting for everyone.
+  - Side steps useless arguments in pull requests.
+- Test with Tasty using Hspec
+  - Tasty is a testing framework for running different kinds of tests.
+  - Hspec tests are the nicest to write.
+  - Use other providers like QuickCheck when they make sense.
+- Benchmark with Criterion
+  - Only if your package needs to be fast!
+  - Otherwise this is a lot of overhead to maintain.
+  - Criterion is the gold standard for pretty much any language.
+- Run tests on Travis CI
```

Comment:
How about links to:

* https://docs.haskellstack.org/en/stable/travis_ci/
* http://www.snoyman.com/blog/2016/08/appveyor-haskell-windows-ci